### ISSUE TITLE
[Feature request] Add "forward" mapping

### ISSUE THREAD

**Issue by geimer** [include-what-you-use/include-what-you-use#988]
The currently available `symbol` mapping already allows to map symbol _definitions_ to header files.  However, in various use cases (see e.g., #530) it is desirable to use an include file for _forward declarations_ rather than explicitly repeating them all over the place.  Unless I have missed something, this doesn't seem to be possible with the `symbol` mapping.  Thus, I propose to implement a `forward` mapping type to also allow such kinds of mappings.

An alternative way of implementing this feature would be to introduce a `forward` visibility type for the `symbol` mapping, though I can't really judge which option makes more sense.

**Comment by kimgr** [comment]
Yep, something like that would be nice.

I like the idea of a `forward` visibility, but my mind is too cloudy to play out exactly how it would work :-)

I'm mostly typing this up for myself, so I can reason about mappings, I'm not entirely sure this is how it works in practice...

For headers:
* Private-to-private mapping says: "From-header includes to-header, but they are both private"
* Private-to-public mapping says: "From-header is private, to-header is public."
* Public-to-public mapping says: "From-header includes to-header, and they are both public"
* Public-to-private mapping should never exist

The outcome of that is that given a suggested header A;
* if A is public and present, leave it
* if A is public and not present, is there a set of public-to-public mapping from A?
    * for every to-header X mapped from A, is X present? If so, leave X.
    * if not, suggest A
* if A is private, whether present or not
    * if there is a private-to-public mapping from A to X, suggest X
    * if not, is there a private-to-private mapping from A to Y? If yes, recurse with Y to see if we can find a public header.
    * if no mapping available, suggest A as-is

For symbols, they are just folded into the mapping visibility structure for convenience. The from-visibility is ignored.
* to-private mapping says: "Symbol is defined in to-header, but it is private"
* to-public mapping says: "Symbol is defined in to-header and it is public."

In conjunction with header mappings this can resolve quite nicely. If a symbol is mapped to a private header and private header mapping eventually gives a public header, that strikes a nice balance between ease of generating automated mappings, e.g. `grep` for symbols and map them to their containing file, then `grep` for `#include` directives and generate header mappings. In practice it's not really that simple, of course.

So I wonder if it would be possible to smuggle a `forward-only` visibility into the unused from-visibility field of the symbol mapping... That would still allow to-private and to-public mappings for symbols, but resolution would have to start by deciding whether the symbol we're looking for is for a forward-declare use or a full use. Something like:

Given a used symbol A;
* if it's a forward-declare use, look up mappings for (A, forward-only) and perform resolution as before
* if it's a full use, look up mappings for (A, !forward-only) and perform resolution as before

So rather than tangling that into each mapping, it might make sense to have a separate database of forward-only mappings for symbols. External mapping files could still express it as `{ symbol: [ "std::cout", "forward-only", "<iosfwd>", "public" ] }`, for example, but they could be separated into a dedicated forward-only map when parsing/loading.

I don't know, it doesn't seem very principled, but maybe it doesn't have to be :).

It would be nice with a more structured definition language for mappings and externalization of pragmas. I've started thinking about that a few times, but never finished anything up.

**Comment by geimer** [comment]
From an outside perspective (i.e., not knowing anything about the implementation), what you wrote up makes sense to me.

If the visibility field is the way to go, maybe one could also consider replacing the `private` symbol visibility keyword by `definition` or similar, while deprecating but still supporting `private` as an alias for backward compatibility for a couple of releases (at least at the level of the mapping file format; what it resolves to internally during parsing is a different matter).  Don't know how much refactoring this would require, though.  Just an idea...

**Comment by bolshakov-a** [comment]
What is the desired behavior of the proposed mapping? Should IWYU replace a fwd-declaration appearing in the code with the mapped header? Or the presence of included header just suppresses the fwd-decl suggestion? As an imaginable use case of the latter option, consider a header `get_class.h` declaring a function `Class GetClass();`. It is known to provide a declaration of `Class`, so may be denoted as providing a fwd-decl of `Class` for other headers which include `get_class.h`. But in other headers, explicit `Class` forward-declarations may still be preferable.

**Comment by bolshakov-a** [comment]
A draft implementation is in #2028.

**Comment by bolshakov-a** [comment]
Implemented in https://github.com/include-what-you-use/include-what-you-use/commit/cc406a285cc30168fc8115de2106e6b6bbfdc943 on master.

**Comment by justusranvier** [comment]
This is great. We've been sorely missing this feature for a long time. With the current version you really have to fight with the tool to get it to consistently use headers that export forward declarations.

Will these mapping directive work with symbols declared in an inline namespace? iwyu refuses to forward declare those types of symbols but it would be great if this behavior could be overridden.

docs/IWYUMappings.md has says "Unlike `include`, `symbol` directives do not support the @-prefixed regex syntax in the first entry. Track the https://github.com/include-what-you-use/include-what-you-use/issues/233 for updates." however that issue is closed. Do regexes work with `symbol` mappings now?

The use case for this would be something like this:

```
{ "symbol": ["@my::proj::a::.*", "fwd-decl", "<my/proj/a/fwd.hpp>", "public"] },
{ "symbol": ["@my::proj::b::.*", "fwd-decl", "<my/proj/b/fwd.hpp>", "public"] },
```

**Comment by bolshakov-a** [comment]
> Will these mapping directive work with symbols declared in an inline namespace?

Good catch! They don't work because fwd-decl uses of classes inside inline namespaces are recategorized to full uses unconditionally:
https://github.com/include-what-you-use/include-what-you-use/blob/673cc339705c828a681a00b06b20bfefc064f3e4/iwyu_output.cc#L842-L846

Moreover, it prevents the IWYU builtin symbol mapping from working correctly with `<iosfwd>` from libc++:
```console
$ include-what-you-use -stdlib=libc++ -xc++ -
#include <iosfwd>       

std::basic_ostream<char, std::char_traits<char>>* os;

<stdin> should add these lines:
#include <ostream>  // for basic_ostream, char_traits

<stdin> should remove these lines:
- #include <iosfwd>  // lines 1-1

The full include-list for <stdin>:
#include <ostream>  // for basic_ostream, char_traits
```
because libc++ uses an inline namespace inside `std`.

> iwyu refuses to forward declare those types of symbols but it would be great if this behavior could be overridden.

Have I understood correctly that you want not only to use fwd-decl headers for names from inline namespaces but also to be able to forward-declare them in your source files? It was decided in #713 that it is rather harmful: https://github.com/include-what-you-use/include-what-you-use/issues/713#issuecomment-549029516. (However, I wonder why #713 appeared. Names starting with `std::` were already recategorized to full uses.)

> Do regexes work with symbol mappings now?

I think they don't, but @kimgr should know better.

**Comment by justusranvier** [comment]
> Have I understood correctly that you want not only to use fwd-decl headers for names from inline namespaces but also to be able to forward-declare them in your source files?

I want to be able to put signifigant portions up to and including the entirety of a project into an inline namespace without iwyu deciding to promote all forward declarations to full uses. I understand this behavior might not be a good default for everyone but I would like a way to achieve that for the symbols I specify.

Consider a namespace strategy inspired from this approach: https://www.foonathan.net/2018/11/inline-namespaces/

My entire project might be declared in:

```
namespace my_library::inline v1
{
...
}
```

Indepedently of whether or not I can create forward declaration headers that Just Work it would be great to not have to fight the tool so much by littering the code with pragmas.

**Comment by kimgr** [comment]
@justusranvier Could you break out a new issue for the inline namespace thing? It looks like I gave up and just forbid fwd-decls for types in inline namespaces, but if there's something smarter we could do, let's try it!

**Comment by kimgr** [comment]
I removed the link to #233 from the docs in 35d09e4fd4e80845510fbf53226ffbb2c88fc338.
